<a href="https://colab.research.google.com/github/Shoaib-Siddiq/L/blob/main/Day_4_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Implement a Multiple Linear Regression model using scikit-learn on a dataset (e.g., California Housing or car price), split the data, train the model, make predictions, and evaluate its performance using MAE, MSE, and R-squared.

## Load the dataset

### Subtask:
Load a suitable dataset (e.g., California Housing or car price) into a pandas DataFrame.


**Reasoning**:
Import necessary libraries and load the California Housing dataset into a pandas DataFrame.



In [1]:
import pandas as pd
from sklearn.datasets import fetch_california_housing

# Load the California Housing dataset
california_housing = fetch_california_housing(as_frame=True)

# Create a DataFrame
df = california_housing.frame

# Display the first few rows and information
display(df.head())
display(df.info())

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  MedHouseVal  
0    -122.23        4.526  
1    -122.22        3.585  
2    -122.24        3.521  
3    -122.25        3.413  
4    -122.25        3.422

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   MedInc       20640 non-null  float64
 1   HouseAge     20640 non-null  float64
 2   AveRooms     20640 non-null  float64
 3   AveBedrms    20640 non-null  float64
 4   Population   20640 non-null  float64
 5   AveOccup     20640 non-null  float64
 6   Latitude     20640 non-null  float64
 7   Longitude    20640 non-null  float64
 8   MedHouseVal  20640 non-null  float64
dtypes: float64(9)
memory usage: 1.4 MB


None

## Prepare the data

### Subtask:
Separate features (X) and target variable (y). Handle any missing values or categorical features if necessary.


**Reasoning**:
Separate the features and target variable and check for missing values.



In [2]:
X = df.drop('MedHouseVal', axis=1)
y = df['MedHouseVal']

print("Missing values in X:")
display(X.isnull().sum())

print("\nMissing values in y:")
display(y.isnull().sum())

Missing values in X:


MedInc        0
HouseAge      0
AveRooms      0
AveBedrms     0
Population    0
AveOccup      0
Latitude      0
Longitude     0
dtype: int64


Missing values in y:


np.int64(0)

## Split the data

### Subtask:
Split the dataset into training and testing sets.


**Reasoning**:
Import the necessary function for splitting the data and then split the features and target into training and testing sets with a specified test size and random state.



In [3]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)

Shape of X_train: (16512, 8)
Shape of X_test: (4128, 8)
Shape of y_train: (16512,)
Shape of y_test: (4128,)


## Build and train the model

### Subtask:
Create a Multiple Linear Regression model using scikit-learn and train it on the training data.


**Reasoning**:
Create and train a Linear Regression model using the training data.



In [4]:
from sklearn.linear_model import LinearRegression

# Instantiate a LinearRegression object
model = LinearRegression()

# Fit the linear regression model to the training data
model.fit(X_train, y_train)

LinearRegression()

## Make predictions

### Subtask:
Use the trained model to make predictions on the testing data.


**Reasoning**:
Use the trained Linear Regression model to make predictions on the test set.



In [5]:
y_pred = model.predict(X_test)

## Evaluate the model

### Subtask:
Calculate and display the Mean Absolute Error (MAE), Mean Squared Error (MSE), and R-squared score to evaluate the model's performance.


**Reasoning**:
Calculate and display the evaluation metrics (MAE, MSE, R-squared) using the actual and predicted values.



In [6]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

# Calculate evaluation metrics
mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

# Print the metrics
print(f"Mean Absolute Error (MAE): {mae}")
print(f"Mean Squared Error (MSE): {mse}")
print(f"R-squared (R2): {r2}")

Mean Absolute Error (MAE): 0.5332001304956553
Mean Squared Error (MSE): 0.5558915986952444
R-squared (R2): 0.5757877060324508


## Summary:

### Data Analysis Key Findings

*   The California Housing dataset, containing 20640 entries and 9 columns (8 features and 1 target), was successfully loaded with no missing values.
*   The data was split into training and testing sets, with 80% (16512 samples) used for training and 20% (4128 samples) for testing.
*   A Multiple Linear Regression model was trained on the training data.
*   The trained model made predictions on the testing data.
*   The model's performance was evaluated using MAE, MSE, and R-squared metrics:
    *   Mean Absolute Error (MAE): 0.5332
    *   Mean Squared Error (MSE): 0.5559
    *   R-squared (R2): 0.5758

### Insights or Next Steps

*   The R-squared value of 0.5758 indicates that approximately 57.6% of the variance in the median house value is explained by the features in the model. This suggests that while the model has some predictive power, there is still a significant portion of the variance not captured.
*   Further steps could involve feature engineering, exploring polynomial features, or trying other regression models (e.g., Ridge, Lasso, or more complex models like Random Forest Regressor) to potentially improve the R-squared score and reduce error metrics.
